docs/guides/plugin_dev/custom_component.rst

```diff
@@ -2,3 +2,77 @@
 
 Custom components
 =================
+
+Components are a large part of how HydroMT defines its behavior.
+A component is basically a part of the model architecture that is responsible
+for handling a specific part of the data, such as the Grid, Config, or Mesh
+but also potentially something like rivers, catchments, pipes or other custom behavior
+your plugin might need.
+
+
+Implementing a component
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Initialisation
+--------------
+
+There are generally two types of components you might want to implement:
+`ModelComponent` and `SpatialModelComponent`. They are similar but the
+`SpatialModelComponent` is meant to hold spacial data. A model MUST have at least one
+`SpatialModelComponent` for it's region functionality to function properly.
+
+Components of any kind, should take a reference to the model they are a part of at
+initialization so that components can access other components through the model as is
+necessary. Typically this is done like so:
+
+.. code-block:: python
+
+    def __init__(
+        self,
+        model: Model,
+        filename: str = "component/{name}.csv",
+    ):
+        self._data: Optional[Dict[str, Union[pd.DataFrame, pd.Series]]] = None
+        self._filename: str = filename
+        super().__init__(model=model)
```

Comment:
Have you tried [doctest](https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html) for these code snippets?